### PR TITLE
feat: Project Tree

### PR DESCRIPTION
**Pull request by hrwX**
Depends on [frappe/295](https://github.com/Bloomstack/frappe/pull/295)

> Please provide enough information so that others can review your pull request:

> Explain the **details** for making this change. What existing problem does the pull request solve?

> Screenshots/GIFs

### CODE REVIEW

**Review by sahil28297 — DISMISSED** [review]

**Review by neilLasrado — DISMISSED** [review]

File: erpnext/public/js/projects/project_tree.js [hunk 1]
```diff
@@ -0,0 +1,1247 @@
...
+				if (strip_html_required) {
+					_value = strip_html(value);
+				} else {
+					_value = typeof value === 'string' ? frappe.utils.escape_html(value) : value;
+				}
+			}
+
+			if (df.fieldtype === 'Image') {
+				html = df.options ?
+					`<img src="${doc[df.options]}" style="max-level: 30px; max-width: 100%;">` :
```
Comment (neilLasrado) [inline erpnext/public/js/projects/project_tree.js]: @DeveloperMetal would it be ok to add inline CSS here or should this be on a separate file?

Comment (DeveloperMetal) [reply erpnext/public/js/projects/project_tree.js]: I would prefer we do not inline any css(unless it depends on dynamic values). that's just dirty and prevents theming from just css files.

File: erpnext/public/js/projects/project_tree.js [hunk 2]
```diff
@@ -0,0 +1,1247 @@
...
+		const df = col.df || {};
+		const label = df.label;
+		const fieldname = df.fieldname;
+		const value = doc[fieldname] || '';
+
+		const format = () => {
+			if (df.fieldtype === 'Code') {
+				return value;
+			} else if (df.fieldtype === 'Percent') {
+				return `<div class="progress level" style="margin: 0px;">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 3]
```diff
@@ -0,0 +1,1247 @@
...
+
+		const liked_by = JSON.parse(doc._liked_by || '[]');
+		let heart_class = liked_by.includes(user) ? 'liked-by' : 'text-extra-muted not-liked';
+
+		const seen = JSON.parse(doc._seen || '[]').includes(user) ? '' : 'bold';
+
+		const subject_link = this.get_subject_link(doc, subject, escaped_subject);
+
+		let html = doc.doctype == 'Task' && doc.expandable ? `<a class="btn btn-action btn-xs"
+			data-doctype="Task" data-name="${escape(doc.name)}" style="width: 20px;">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 4]
```diff
@@ -0,0 +1,1247 @@
...
+				<i class="octicon octicon-chevron-right" />
+			</a>` : ``;
+
+
+		let subject_html = `
+			<input class="level-item list-row-checkbox hidden-xs" type="checkbox" data-name="${escape(doc.name)}">
+			<span class="level-item" style="margin-bottom: 1px;">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 5]
```diff
@@ -0,0 +1,1247 @@
...
+				<i class="octicon octicon-heart like-action ${heart_class}"
+					data-name="${doc.name}" data-doctype="${doc.doctype}"
+					data-liked-by="${encodeURI(doc._liked_by) || '[]'}">
+				</i>
+				<span class="likes-count">
+					${ liked_by.length > 99 ? __("99") + '+' : __(liked_by.length || '')}
+				</span>
+			</span>
+			<span class="level-item ${seen} ellipsis" title="${escaped_subject}" style="padding-left: ${20*level}px;">
+				<span class="level-item" style="margin-bottom: 1px;"">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 6]
```diff
@@ -0,0 +1,1247 @@
...
+				${this.get_indicator_dot(doc)}
+			</div>
+		`;
+
+		return html;
+	}
+
+	get_open_button(doc) {
+		return `
+				<div class="level-item hidden-xs" style="margin-left: 5px;">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 7]
```diff
@@ -0,0 +1,1247 @@
...
+
+		return this.get_header_html_skeleton($columns, '<span class="list-count"></span>');
+	}
+
+	get_previous_header_html() {
+		return `
+			<header class="level list-row list-row-head text-muted small">
+				<a class="btn btn-prev btn-xs">
+					<i class="octicon octicon-chevron-left" />
+					<span style="margin-left: 5px">Projects</span>
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 8]
```diff
@@ -0,0 +1,1247 @@
...
+				</a>
+				<button class="btn btn-xs expand-all btn-default" style="float: right">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

File: erpnext/public/js/projects/project_tree.js [hunk 9]
```diff
@@ -0,0 +1,1247 @@
...
+					${__('Expand All')}</button>
+				<button class="btn btn-xs collapse-all btn-default" style="float: right; display: none">
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js]: No CSS Inlining please

**Review by DeveloperMetal — CHANGES_REQUESTED** [review]
Only use css inlining if you are setting values dynamically please. Or a modular css lib but that's kinda late for this project.

**Review by neilLasrado — DISMISSED** [review]

**Comment by neilLasrado** [comment]
@DeveloperMetal Can you review this again ?

File: erpnext/public/js/projects/project_tree.js [hunk 10]
```diff
@@ -0,0 +1,1247 @@
...
+			let project = unescape(e.currentTarget.getAttribute("data-project"));
+			let task = frappe.model.get_new_doc("Task");
+			task["project"] = project;
+			task["parent_task"] = parent;
+
+			frappe.ui.form.make_quick_entry("Task", null, null, task);
+		});
+	}
+
+	get_projects() {
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js:482]: Lets rename this to `setup_get_project_events()` to stay consistent with your other event setup naming conventions

File: erpnext/public/js/projects/project_tree.js [hunk 11]
```diff
@@ -0,0 +1,1247 @@
...
+		frappe.call(this.get_call_args([])).then(r => {
+			// render
+			this.render_header(this.columns, true);
+			this.prepare_data(r);
+			this.toggle_result_area();
+			this.render();
+		});
+	}
+
+	get_tasks() {
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js:503]: Lets rename this to `setup_get_task_events()` to stay consistent with your other event setup naming conventions

File: erpnext/public/js/projects/project_tree.js [hunk 12]
```diff
@@ -0,0 +1,1247 @@
...
+	get_left_html(columns, doc, level) {
+		return columns.map(col => this.get_column_html(col, columns, doc, level)).join('');
+	}
+
+	get_right_html(doc, create_new=false) {
+		return this.get_meta_html(doc, create_new);
+	}
+
+	get_column_html(col, columns, doc, level) {
+		if (col.type === 'Status') {
```
Comment (DeveloperMetal) [inline erpnext/public/js/projects/project_tree.js:678]: This is a pretty cool setup of using code, percent or format fields. You should really add some descriptions here as how to use them and when.

File: erpnext/accounts/doctype/subscription/subscription.py [hunk 13]
```diff
@@ -469,13 +469,16 @@ def process_for_active(self):
 		3. Change the `Subscription` status to 'Cancelled'
 		"""
 		if getdate() > getdate(self.current_invoice_end) and self.is_prepaid_to_invoice():
+			print("1")
```
Comment (neilLasrado) [inline erpnext/accounts/doctype/subscription/subscription.py:473]: You need to remove the prints

**Review by neilLasrado — CHANGES_REQUESTED** [review]

**Pull request closed without merging**